core-clp: Add support for decompressing a specific file split from a clp archive into one or more IR files.

haiqi96 opened the pull request:
# References

# Description
The change is motivated by the need to support log viewer, which 
1. Requires the input to be in the IR format.
2. Requires the IR to be reasonably small to not overflow the memory.

This PR introduces a new decompression interface `decompress_ir` that decompress a file split into one or multiple IRs.

The function takes in an original file ID, a specific message index and a threshold. 
It first find the file split which contains the message index, and decompress the split into one or more IR; the function creates a new IR whenever the current IR's raw size (i.e. not zstd compressed) is greater than the given threshold.
Each IR follows the IRv1 format, meaning it has the complete preamble and and EoF byte, and can be deserialized individually.

The generated IR use the naming format: <FILE_ORIG_ID>_<begin_message_ix>_<end_message_ix>.clp.zst. Since the preamble of the IRv1 doesn't contain any log event index information, this name is essential for the user of the IR to know what's the range of log index the IR contains.

The PR also introduces a new class `LogEventSerializer.cpp` that serialized a plain text message into the IR format.

Due to the limitation of our current IR related encoding APIs, the function is designed with two inefficienies
1. The serialized data is first written into an internal buffer, instead of directly writing to the on-disk file.
2. The serializer takes in a plain text message, meaning that the encoded message in the archive needs to be first decompressed into plain text (which introduce extra overhead) before being serialized.

We agreed that these two are acceptables as properly supporting the flow will take more thoughts on reworking the encoder interface.

# Validation performed
The validation is not directly performed on this PR, but on a following PR which adds the  `decompress_ir` in to the execution path of clp executable.

To validate the functionality, we compressed a 64MB file into archive(s). We then decompressed it into mulitple IRs, decoded and concatnate them,  and did a binary comparison with the original file. 

We used two configuration to cover all the possible cases:

1. Compressed a 64MB hadoop log using smaller encoded file size and archive size, such that it splits the original file into 3 splits across 2 archives. We then decompressed all 3 IRs by running clp 3 times, using different message index

2. Compressed the 64MB hadoop log using default settings, so only one file and archive was generated. We then decompressed the IR using a 32MB threshold, generating 3 IRs on disk.

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+#include "LogEventSerializer.hpp"
+
+#include <string_utils/string_utils.hpp>
+
+#include "../ffi/ir_stream/encoding_methods.hpp"
+#include "../ffi/ir_stream/protocol_constants.hpp"
+#include "spdlog_with_specializations.hpp"
```
```suggestion
#include <string>
#include <string_view>

#include <spdlog/spdlog.h>

#include "../ErrorCode.hpp"
#include "../ffi/ir_stream/encoding_methods.hpp"
#include "../ffi/ir_stream/protocol_constants.hpp"
#include "../type_utils.hpp"
#include "types.hpp"
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+    if (false == m_is_open) {
+        throw OperationFailed(ErrorCode_Unsupported, __FILENAME__, __LINE__);
+    }
+    m_zstd_compressor.write(reinterpret_cast<char const*>(m_ir_buffer.data()), m_ir_buffer.size());
```
```suggestion
    m_zstd_compressor.write(
            size_checked_pointer_cast<char const>(m_ir_buffer.data()),
            m_ir_buffer.size()
    );
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    string logtype;
+    bool res;
```
```suggestion
    bool res{};
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+                m_ir_buffer
+        );
+    } else {
+        auto timestamp_delta = timestamp - m_prev_msg_timestamp;
```
```suggestion
        auto const timestamp_delta = timestamp - m_prev_msg_timestamp;
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+
+// Explicitly declare template specializations so that we can define the template methods in this
+// file
+template LogEventSerializer<eight_byte_encoded_variable_t>::LogEventSerializer();
+template LogEventSerializer<four_byte_encoded_variable_t>::LogEventSerializer();
```
```suggestion
```
We don't need this specialization since they are not implemented in this file

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+}
+
+template <typename encoded_variable_t>
+auto LogEventSerializer<encoded_variable_t>::flush() -> void {
```
this function should come before `close` to match the ordering in the header file

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+        epoch_time_ms_t timestamp
+) -> ErrorCode {
+    if (false == m_is_open) {
+        throw OperationFailed(ErrorCode_Unsupported, __FILENAME__, __LINE__);
```
```suggestion
        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+template <typename encoded_variable_t>
+auto LogEventSerializer<encoded_variable_t>::flush() -> void {
+    if (false == m_is_open) {
+        throw OperationFailed(ErrorCode_Unsupported, __FILENAME__, __LINE__);
```
```suggestion
        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+namespace clp::ir {
+template <typename encoded_variable_t>
+LogEventSerializer<encoded_variable_t>::~LogEventSerializer() {
+    if (true == m_is_open) {
```
```suggestion
    if (m_is_open) {
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,168 @@
+auto LogEventSerializer<encoded_variable_t>::close() -> void {
+    m_ir_buffer.push_back(clp::ffi::ir_stream::cProtocol::Eof);
+    flush();
+    m_zstd_compressor.close();
+    m_writer.close();
```
This is trivial, but I think we could make a helper to wrap these two methods so we never mess up the order for closing them

LinZhihao-723 reviewed (changes requested):
Add review comments in LogEventSerializer.hpp/cpp

LinZhihao-723 commented on components/core/src/clp/ir/utils.hpp:
```diff
@@ -3,12 +3,31 @@
  * @return Whether the content in the buffer starts with one of the IR stream magic numbers
  */
 auto has_ir_stream_magic_number(std::string_view buf) -> bool;
+/**
+ * @param buf
+ * @return Extension name of IR
+ */
+auto get_ir_extension_name() -> std::string;
```
Do we need this method? Can we just create a constant?

haiqi96 replied:
Sure, it works. it looks like I should create a new Constant.hpp file under ir/ directory, or do you have any suggestion?

LinZhihao-723 replied:
I don't have any preference. @kirkrodrigues shall we create a new file?

LinZhihao-723 commented on components/core/src/clp/ir/utils.hpp:
We should add an empty line to separate each function

LinZhihao-723 commented on components/core/src/clp/ir/utils.hpp:
```diff
@@ -3,12 +3,31 @@
+ * @return Extension name of IR
+ */
+auto get_ir_extension_name() -> std::string;
+/**
+ * Estimate the approximate upper bound of the given log message when encoded as IR.
+ * This function makes the following assumption to for a rough upper bound estimation
+ * 1. dictionary variable lengths are all encoded with int32_t
+ * 2. timestamp or timestamp delta is encoded with int64_t
+ * 3. log type length is encoded with int32_t
+ * 4. The size of encoded variable roughly equals to their plain text size.
+ * @param log_message
+ * @param num_encoded_vars
+ * @return the approximated ir size in bytes
+ */
```
```suggestion
/**
 * Gets an approximated upper bound size of a given log message if it is encoded in IR format.
 * The approximation is based on the following assumptions:
 * 1. Dictionary variable lengths are all encoded using int32_t
 * 2. Timestamp or timestamp delta is encoded using int64_t
 * 3. Logtype length is encoded using int32_t
 * 4. The size of encoded variable roughly equals to their plain text size.
 * @param log_message
 * @param num_encoded_vars
 * @return the approximated ir size in bytes
 */
```

LinZhihao-723 commented on components/core/src/clp/ir/utils.cpp:
```diff
@@ -2,6 +2,7 @@
 
```
```suggestion

#include <cstddef>
#include <string_view>

```

LinZhihao-723 commented on components/core/src/clp/ir/utils.cpp:
```diff
@@ -10,4 +11,35 @@ auto has_ir_stream_magic_number(std::string_view buf) -> bool {
 }
+
+auto get_approximated_ir_size(std::string_view log_message, size_t num_encoded_vars) -> size_t {
+    size_t const tag_size = sizeof(char);
+
+    // sizeof(log type) + sizeof (dict vars) + sizeof(encoded_vars) ~= The size of log message
+    auto ir_size = log_message.size();
+
+    // Add the size of placeholders in the original log type
+    ir_size += num_encoded_vars * sizeof(enum_to_underlying_type(VariablePlaceholder()));
+
+    // Add the tags and encoding length bytes of log type
+    size_t const logtype_length_encoding_size = sizeof(int32_t);
+    ir_size += tag_size + logtype_length_encoding_size;
+
+    // Add the tags and encoding length bytes for dictionary variables
+    // Note here we overestimate the size by assuming that encoded float
+    // and int also have length encoding bytes
+    size_t const var_encoding_size = sizeof(int32_t);
+    ir_size += (tag_size + var_encoding_size) * num_encoded_vars;
+
+    // Add the tags and encoding length bytes of log type
+    size_t const timestamp_length_encoding_size = sizeof(int64_t);
+    ir_size += tag_size + timestamp_length_encoding_size;
+
+    return ir_size;
```
```suggestion
    constexpr size_t cLogtypeLengthSize = sizeof(int32_t);
    constexpr size_t cTagSize = sizeof(char);
    constexpr size_t cVarSize = sizeof(int32_t);
    constexpr size_t cTimestampSize = sizeof(int64_t);
    constexpr size_t cPlaceHolderSize = sizeof(enum_to_underlying_type(VariablePlaceholder()));

    // sizeof(log type) + sizeof (dict vars) + sizeof(encoded_vars) ~= The size of log message
    auto ir_size = log_message.size();

    // Add the size of placeholders in the original log type
    ir_size += num_encoded_vars * cPlaceHolderSize;

    // Add the tags and encoding length bytes of log type
    ir_size += cTagSize + cLogtypeLengthSize;

    // Add the tags and encoding length bytes for dictionary variables
    // Note here we overestimate the size by assuming that encoded float and int also have length
    // encoding bytes
    ir_size += (cTagSize + cVarSize) * num_encoded_vars;

    // Add the tags and encoding length bytes of log type
    ir_size += cTagSize + cTimestampSize;

    return ir_size;
```

LinZhihao-723 commented on components/core/src/clp/ir/utils.hpp:
```diff
@@ -3,12 +3,31 @@
+ */
+auto get_ir_extension_name() -> std::string;
+/**
+ * Estimate the approximate upper bound of the given log message when encoded as IR.
+ * This function makes the following assumption to for a rough upper bound estimation
+ * 1. dictionary variable lengths are all encoded with int32_t
+ * 2. timestamp or timestamp delta is encoded with int64_t
+ * 3. log type length is encoded with int32_t
+ * 4. The size of encoded variable roughly equals to their plain text size.
```
```suggestion
 * Gets an approximated upper bound size of a given log message if it is encoded in IR format.
 * The approximation is based on the following assumptions:
 * 1. Dictionary variable lengths are all encoded using int32_t
 * 2. Timestamp or timestamp delta is encoded using int64_t
 * 3. Logtype length is encoded using int32_t
 * 4. The size of encoded variable roughly equals to their plain text size.
```

LinZhihao-723 reviewed (changes requested):
Left some comments on ir/utils.hpp/cpp. However, I'm thinking whether this approximation function is necessary for the following reasons:
1. During the compression, it is called before we serialize a message to determine whether to start a rotate to a new IR file. But do we really need this precision? Can we serialize the message anyway and in the end of a serialization call we check if the total size exceeds the threshold? Unless we have a concrete use case where post-check makes a huge difference to the pre-check, I don't think we need to push the effort. We are compressing the stream anyways, so the actual bytes written to the disk is not what we count.
2. This API is not generic: it is assuming you already know the number of encoded variables, plus the message string. But normally you can't get the number of encoded variables if you don't execute message encoding. But if you already execute encoding, you don't need the estimation. If we ended up keeping this method, we should move it to where it's used and make it as a local helper.

LinZhihao-723 commented on components/core/src/clp/streaming_archive/reader/Archive.hpp:
```diff
@@ -99,6 +99,15 @@ class Archive {
     decompress_message(File& file, Message const& compressed_msg, std::string& decompressed_msg);
 
+    /**
+     * Decompresses a given message from a given file without its original timestamp
```
```suggestion
     * Decompresses a given message from a given file without formatting its timestamp
```

LinZhihao-723 commented on components/core/src/clp/streaming_archive/reader/Archive.hpp:
```diff
@@ -99,6 +99,15 @@ class Archive {
+     * Decompresses a given message from a given file without its original timestamp
+     * @param compressed_msg
+     * @param decompressed_msg
+     * @return true if message was successfully decompressed, false otherwise
```
```suggestion
     * @return Whether the message was successfully decompressed
```

LinZhihao-723 commented on components/core/src/clp/streaming_archive/reader/Archive.cpp:
```diff
@@ -216,6 +200,33 @@ bool Archive::decompress_message(
+    decompressed_msg.clear();
+
+    // Build original message content
+    logtype_dictionary_id_t const logtype_id = compressed_msg.get_logtype_id();
```
```suggestion
    auto const logtype_id = compressed_msg.get_logtype_id();
```

LinZhihao-723 commented on components/core/src/clp/streaming_archive/reader/Archive.cpp:
```diff
@@ -216,6 +200,33 @@ bool Archive::decompress_message(
+    // Build original message content
+    logtype_dictionary_id_t const logtype_id = compressed_msg.get_logtype_id();
+    auto const& logtype_entry = m_logtype_dictionary.get_entry(logtype_id);
+    if (!EncodedVariableInterpreter::decode_variables_into_message(
```
```suggestion
    if (false == EncodedVariableInterpreter::decode_variables_into_message(
```

LinZhihao-723 reviewed (changes requested):
nit refactoring. Since we're moving the code around let's rewrite it using the latest standard

haiqi96 commented on components/core/src/clp/ir/LogEventSerializer.hpp:
```diff
@@ -0,0 +1,117 @@
+     */
+    auto open(std::string const& file_path) -> ErrorCode;
+
+    /**
+     * Creates a Zstandard compressed four bytes encoded IR on the disk, and writes the preamble to
+     * the IR
+     * @param file_path
+     * @param epoch_time_ms_t
+     * @return
+     */
+    auto open(std::string const& file_path, epoch_time_ms_t ref_timestamp) -> ErrorCode;
```
what you suggested was my initial proposal, but Kirk suggested the current one.
We do have a static assert in the code, so if someone is using the wrong open, compiler should catch it

kirkrodrigues replied:
Sure, we can go with the previous proposal.

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+    }
+
+    boost::filesystem::path temp_ir_path{temp_output_dir};
+    temp_ir_path /= m_encoded_file.get_id_as_string() + ".temp." + ir::get_ir_extension_name();
```
Do you think the extension name ".clp.zst" makes more sense?

haiqi96 replied:
you mean remove the ".temp."? sure

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+                        end_message_ix
+                ))
+            {
+                return false;
```
A general question: when failure happens, shall we close the serializer before return?

haiqi96 replied:
technically we should, however, we are not closing archive or metadata either. another issue I can see: if a serializing a log event fails, the IR will be incompletely anyway.

For this specific use case, since we are only decompressing a single file, I don't think closing or not closing makes a difference.
If one day we are going to decompress multiple IRs in a single execution, we can further optimize this

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -1,13 +1,55 @@
+ * @param end_message_ix
+ * @return true if the IR is renamed and moved successfully. Otherwise false
+ */
+bool rename_and_move_ir(
```
"rename" and "move" here should have the same meaning. We should call it either "rename_ir_file" or "move_ir_file"

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -1,13 +1,55 @@
+    std::string ir_file_name = file_orig_id + "_" + std::to_string(begin_message_ix) + "_"
+                               + std::to_string(end_message_ix) + "." + ir::get_ir_extension_name();
+
+    auto renamed_ir_path = output_directory / ir_file_name;
```
```suggestion
    auto const renamed_ir_path = output_directory / ir_file_name;
```

LinZhihao-723 replied:
Shall we check whether the destination path already exists?

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+        size_t ir_target_size
+) {
+    // Open encoded file
+    auto error_code = archive_reader.open_file(m_encoded_file, file_metadata_ix);
+    if (ErrorCode_Success != error_code) {
```
```suggestion
    if (auto const error_code = archive_reader.open_file(m_encoded_file, file_metadata_ix);
        ErrorCode_Success != error_code)
    {
```

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+    }
+
+    // Generate output directory
+    error_code = create_directory_structure(output_dir, 0700);
+    if (ErrorCode_Success != error_code) {
```
```suggestion
    if (auto const error_code = create_directory_structure(output_dir, 0700);
        ErrorCode_Success != error_code)
    {
```

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+
+    if (temp_output_dir != output_dir) {
+        // Generate temporary output directory
+        error_code = create_directory_structure(temp_output_dir, 0700);
+        if (ErrorCode_Success != error_code) {
```
```suggestion
        if (auto const error_code = create_directory_structure(temp_output_dir, 0700);
            ErrorCode_Success != error_code)
        {
```

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+
+    LogEventSerializer<four_byte_encoded_variable_t> ir_serializer;
+    // Open output IR file
+    error_code = ir_serializer.open(temp_ir_path.string(), m_encoded_file.get_begin_ts());
+    if (ErrorCode_Success != error_code) {
```
```suggestion
    if (auto const error_code
        = ir_serializer.open(temp_ir_path.string(), m_encoded_file.get_begin_ts());
        ErrorCode_Success != error_code)
    {
```

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+    }
+
+    while (archive_reader.get_next_message(m_encoded_file, m_encoded_message)) {
+        if (!archive_reader
+                     .decompress_message_without_ts(m_encoded_message, m_decompressed_message))
```
```suggestion
        if (false
            == archive_reader
                       .decompress_message_without_ts(m_encoded_message, m_decompressed_message))
```

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+        if (message_size_as_ir + ir_serializer.get_serialized_size() > ir_target_size) {
+            ir_serializer.close();
+
+            auto end_message_ix = begin_message_ix + ir_serializer.get_num_log_events();
```
```suggestion
            auto const end_message_ix = begin_message_ix + ir_serializer.get_num_log_events();
```
Also, what do you think to make `end_message_ix = begin_message_ix + ir_serializer.get_num_log_events() - 1`? In this way, begin and end will both be inclusive.

kirkrodrigues replied:
Resolved offline: We will stick with end_message_ix being exclusive.

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+            }
+            begin_message_ix = end_message_ix;
+
+            error_code = ir_serializer.open(
+                    temp_ir_path.string(),
+                    m_encoded_message.get_ts_in_milli()
+            );
+            if (ErrorCode_Success != error_code) {
```
```suggestion
            if (auto const error_code
                = ir_serializer.open(temp_ir_path.string(), m_encoded_message.get_ts_in_milli());
                ErrorCode_Success != error_code)
            {
```

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+            }
+        }
+
+        error_code = ir_serializer.serialize_log_event(
+                m_decompressed_message,
+                m_encoded_message.get_ts_in_milli()
+        );
+        if (ErrorCode_Success != error_code) {
```
```suggestion
        if (auto const error_code = ir_serializer.serialize_log_event(
                    m_decompressed_message,
                    m_encoded_message.get_ts_in_milli()
            );
            ErrorCode_Success != error_code)
        {
```

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+            return false;
+        }
+    }
+    auto end_message_ix = begin_message_ix + ir_serializer.get_num_log_events();
```
```suggestion
    auto const end_message_ix = begin_message_ix + ir_serializer.get_num_log_events();
```

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+    {
+        return false;
+    }
+
```
I can tell the reason why you don't remove `temp_output_dir` (since in the current implementation you don't know if it already exists before this call), but can we add a comment to explain this behavior? It is weird to create some temp resources without cleaning up.

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+                m_encoded_message.get_ts_in_milli()
+        );
+        if (ErrorCode_Success != error_code) {
+            SPDLOG_ERROR("Failed to serialize log event: {}", m_decompressed_message.c_str());
```
If we want to log the log event failed to serialize, we should also log the timestamp

LinZhihao-723 reviewed (changes requested):
Reviewed `FileCompressor`

haiqi96 commented:
> Left some comments on ir/utils.hpp/cpp. However, I'm thinking whether this approximation function is necessary for the following reasons:
> 
> 1. During the compression, it is called before we serialize a message to determine whether to start a rotate to a new IR file. But do we really need this precision? Can we serialize the message anyway and in the end of a serialization call we check if the total size exceeds the threshold? Unless we have a concrete use case where post-check makes a huge difference to the pre-check, I don't think we need to push the effort. We are compressing the stream anyways, so the actual bytes written to the disk is not what we count.
> 2. This API is not generic: it is assuming you already know the number of encoded variables, plus the message string. But normally you can't get the number of encoded variables if you don't execute message encoding. But if you already execute encoding, you don't need the estimation. If we ended up keeping this method, we should move it to where it's used and make it as a local helper.

try reply

LinZhihao-723 commented:
> > Left some comments on ir/utils.hpp/cpp. However, I'm thinking whether this approximation function is necessary for the following reasons:
> > 
> > 1. During the compression, it is called before we serialize a message to determine whether to start a rotate to a new IR file. But do we really need this precision? Can we serialize the message anyway and in the end of a serialization call we check if the total size exceeds the threshold? Unless we have a concrete use case where post-check makes a huge difference to the pre-check, I don't think we need to push the effort. We are compressing the stream anyways, so the actual bytes written to the disk is not what we count.
> > 2. This API is not generic: it is assuming you already know the number of encoded variables, plus the message string. But normally you can't get the number of encoded variables if you don't execute message encoding. But if you already execute encoding, you don't need the estimation. If we ended up keeping this method, we should move it to where it's used and make it as a local helper.
> 
> Answer to the first point. we want to avoid a single huge message overflows the IR. based one what we observed in the past log, this is possible. We have the option to first serialize the message, and manage to "roll back" if the message turns out to be too large, but it will add unnecessary complication to the Serializer class.
> 
> For 2. sure, we can move it to the FileDecompressor's unnamed namespace

For 1: If we look at the logic we used to split files in CLP compression, we split files using existing bytes instead of making approximations. Also, in some scenarios, if the first log event to serialize is larger than the rotation threshold, the current code would first rotate an empty IR, create a new one, and write the large log event into the file (and it is still an overflow). In general, I still advocate for pre-checking. If we want to avoid overflow, we can have some safety factor applied to the IR threshold, for example, multiply the given threshold by 0.9. @kirkrodrigues what do you think?

haiqi96 commented:
> > > Left some comments on ir/utils.hpp/cpp. However, I'm thinking whether this approximation function is necessary for the following reasons:
> > > 
> > > 1. During the compression, it is called before we serialize a message to determine whether to start a rotate to a new IR file. But do we really need this precision? Can we serialize the message anyway and in the end of a serialization call we check if the total size exceeds the threshold? Unless we have a concrete use case where post-check makes a huge difference to the pre-check, I don't think we need to push the effort. We are compressing the stream anyways, so the actual bytes written to the disk is not what we count.
> > > 2. This API is not generic: it is assuming you already know the number of encoded variables, plus the message string. But normally you can't get the number of encoded variables if you don't execute message encoding. But if you already execute encoding, you don't need the estimation. If we ended up keeping this method, we should move it to where it's used and make it as a local helper.
> > 
> > 
> > Answer to the first point. we want to avoid a single huge message overflows the IR. based one what we observed in the past log, this is possible. We have the option to first serialize the message, and manage to "roll back" if the message turns out to be too large, but it will add unnecessary complication to the Serializer class.
> > For 2. sure, we can move it to the FileDecompressor's unnamed namespace
> 
> For 1: If we look at the logic we used to split files in CLP compression, we split files using existing bytes instead of making approximations. Also, in some scenarios, if the first log event to serialize is larger than the rotation threshold, the current code would first rotate an empty IR, create a new one, and write the large log event into the file (and it is still an overflow). In general, I still advocate for pre-checking. If we want to avoid overflow, we can have some safety factor applied to the IR threshold, for example, multiply the given threshold by 0.9. @kirkrodrigues what do you think?

I thought about letting the serializer internally handle rotating the IR, but given that the IR needs to be renamed and moved, it would be tricky to let the serializer notify the caller.
For CLP use case, I believe the memory isn't a strict requirement, but the IR size could be a relatively strict requirement for log viewer.

Would you be clear that whether you don't like the idea of pre-checking, or just don't like the estimation function because it looks complicated? The simplist thing I can think of is to just not use the function, but use the uncompressed message size, which would still be a fair estimation. (as long as the message doesn't contains a tons of non-dictionary variables that only span 1~2 bytes in the original text)

kirkrodrigues commented:
This is rough, but what if we made the serializer handle the splitting internally to some extent:

* Let's say we make the serializer also store a threshold at which it should write no more data.
* When the user appends a log event that causes the threshold to be exceeded, the serializer disables any future writes until the IR stream is rotated.
* To rotate the IR stream, the user needs to call some new API like `rotate(string_view new_file_path)` that will:
  * flush the buffer up to but **not including** the last log event
  * close the current IR stream
  * open and start a new IR stream and write the buffered log event

In theory, we could decouple the serializer from a file so that by default it will buffer everything unless we provide a file for it to flush into. This could potentially allow us to use RAII for the serializer if we rewrite the APIs a bit.

After discussing it with @haiqi96, he prefers to keep things simple so we'll remove the estimation for now and let the user estimate a good threshold themselves; if this becomes a problem in production, we'll have more data that we can use to come up with a better solution.

LinZhihao-723 commented:
> This is rough, but what if we made the serializer handle the splitting internally to some extent:
> 
> * Let's say we make the serializer also store a threshold at which it should write no more data.
> * When the user appends a log event that causes the threshold to be exceeded, the serializer disables any future writes until the IR stream is rotated.
> * To rotate the IR stream, the user needs to call some new API like `rotate(string_view new_file_path)` that will:
>   
>   * flush the buffer up to but **not including** the last log event
>   * close the current IR stream
>   * open and start a new IR stream and write the buffered log event
> 
> In theory, we could decouple the serializer from a file so that by default it will buffer everything unless we provide a file for it to flush into. This could potentially allow us to use RAII for the serializer if we rewrite the APIs a bit.
> 
> After discussing it with @haiqi96, he prefers to keep things simple so we'll remove the estimation for now and let the user estimate a good threshold themselves; if this becomes a problem in production, we'll have more data that we can use to come up with a better solution.

sgtm

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.hpp:
```diff
@@ -0,0 +1,114 @@
+    auto flush() -> void;
+
+    /**
+     * Writes the EoF tag to the end of IR, flushes the data and closes the serializer
```
```suggestion
     * Writes the EoF tag to the end of IR, flushes the data and closes the serializer.
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.hpp:
```diff
@@ -0,0 +1,114 @@
+    [[nodiscard]] auto get_num_log_events() const -> size_t { return m_num_log_events; }
+
+    /**
+     * Serializes a log event and writes it to the end of the internal buffer
```
```suggestion
     * Serializes a log event into the internal buffer.
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.hpp:
```diff
@@ -0,0 +1,114 @@
+    // Method
+    auto close_writer() -> void;
+
+    // Constant
```
Constant should come before methods

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,151 @@
+}
+
+template <typename encoded_variable_t>
+auto LogEventSerializer<encoded_variable_t>::open(string const& file_path) -> ErrorCode {
```
Still need to add a double open check right?

LinZhihao-723 commented on components/core/src/clp/ir/Constant.hpp:
```diff
@@ -0,0 +1,8 @@
+#define CLP_IR_CONSTANTS_HPP
+
+namespace clp::ir {
+constexpr char cIrExtension[] = "clp.zst";
```
I think I wasn't clear in the previous comment (I commented in where this constant was used). I think it makes more sense to make it `.clp.zst`, as `.` should be a part of the extension. Say you want to check if a file is zstd compressed IR, you check if it ends with `.clp.zst` instead of `clp.zst` right? You could have `IamUsingclp.zst` which should not be considered as a clp file, neither with a clp IR extension

haiqi96 replied:
Fair. I wasn't sure about if I need to add ".". After your explanation, adding "." to the extension makes more sense

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,151 @@
+    if (false == res) {
+        return ErrorCode_Failure;
+    }
+    m_num_log_events += 1;
```
```suggestion
    ++m_num_log_events;
```
nit

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -1,13 +1,56 @@
 namespace clp::clp {
+namespace {
+/**
+ * Rename a temporary IR and move it to the output directory
+ * The new name follows the following format
```
```suggestion
 * Renames a temporary IR and moves it to the output directory.
 * The new name follows the following format:
```

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -1,13 +1,56 @@
+ * @param file_orig_id
+ * @param begin_message_ix
+ * @param end_message_ix
+ * @return true if the IR is renamed and moved successfully. Otherwise false
```
```suggestion
 * @return Whether the IR file is successfully renamed.
```

LinZhihao-723 reviewed (dismissed):
Commit message suggestion:
`core-clp: Add support for decompressing an IR from a specific file split from a clp archive.`
I've done my parts of review, maybe you can take it over from here @kirkrodrigues

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.hpp:
```diff
@@ -24,6 +24,14 @@ class FileDecompressor {
+            std::string const& output_dir,
+            std::string const& temp_output_dir,
+            streaming_archive::reader::Archive& archive_reader,
+            size_t ir_target_size
```
include cstddef

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.hpp:
```diff
@@ -24,6 +24,14 @@ class FileDecompressor {
             std::unordered_map<std::string, std::string>& temp_path_to_final_path
     );
 
+    bool decompress_to_ir(
```
Let's add a docstring. I know the other method doesn't have a docstring, but it should.

haiqi96 replied:
As discussed, will add in future changelist

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +119,124 @@ bool FileDecompressor::decompress_file(
+    if (auto const error_code = ir_serializer.open(temp_ir_path.string());
+        ErrorCode_Success != error_code)
+    {
+        return false;
```
If we want to log the failure, this would be the place to do it.

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +119,124 @@ bool FileDecompressor::decompress_file(
+            SPDLOG_ERROR("Failed to decompress message");
+            return false;
+        }
+        if (ir_serializer.get_serialized_size() >= ir_target_size) {
```
```suggestion

        if (ir_serializer.get_serialized_size() >= ir_target_size) {
```

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -1,13 +1,56 @@
 namespace clp::clp {
+namespace {
+/**
+ * Renames a temporary IR and moves it to the output directory.
+ * The new name follows the following format:
+ * <FILE_ORIG_ID>_<begin_message_ix>_<end_message_ix>.clp.zst
```
```suggestion
 * Renames a temporary IR file and moves it to the output directory.
 *
 * The new name uses the following format:
 * <orig_file_id>_<begin_message_ix>_<end_message_ix>.clp.zst
```

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -1,13 +1,56 @@
+ * <FILE_ORIG_ID>_<begin_message_ix>_<end_message_ix>.clp.zst
+ * @param temp_ir
+ * @param output_directory
+ * @param file_orig_id
```
orig_file_id?

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -1,13 +1,56 @@
+bool rename_ir_file(
+        boost::filesystem::path const& temp_ir_path,
+        boost::filesystem::path const& output_directory,
+        std::string const& file_orig_id,
```
```suggestion
        string const& file_orig_id,
```

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -1,13 +1,56 @@
+        size_t begin_message_ix,
+        size_t end_message_ix
+) {
+    std::string ir_file_name = file_orig_id + "_" + std::to_string(begin_message_ix) + "_"
```
```suggestion
    string ir_file_name = file_orig_id + "_" + std::to_string(begin_message_ix) + "_"
```

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +119,124 @@ bool FileDecompressor::decompress_file(
+    auto const end_message_ix = begin_message_ix + ir_serializer.get_num_log_events();
+    ir_serializer.close();
+
+    // Note we don't remove the temp_output_dir because we don't know if it exists before execution
```
```suggestion
    // NOTE: We don't remove temp_output_dir because we don't know if it existed before this method
    // was called.
```

kirkrodrigues commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +119,124 @@ bool FileDecompressor::decompress_file(
+        return false;
+    }
+
+    // Close files
```
```suggestion
```

kirkrodrigues commented on components/core/src/clp/streaming_archive/reader/Archive.hpp:
```diff
@@ -89,7 +89,7 @@ class Archive {
 
     /**
-     * Decompresses a given message from a given file
+     * Decompresses a given message from a given file with the original timestamp
```
```suggestion
     * Decompresses the given message from the given file, including inserting and formatting its
     * timestamp if necessary.
```

kirkrodrigues commented on components/core/src/clp/streaming_archive/reader/Archive.hpp:
```diff
@@ -99,6 +99,15 @@ class Archive {
     decompress_message(File& file, Message const& compressed_msg, std::string& decompressed_msg);
 
+    /**
+     * Decompresses a given message from a given file without appending its original timestamp
```
```suggestion
     * Decompresses the given message without inserting its timestamp.
```

kirkrodrigues commented:
Forgot to mention, since we now have a log event serializer, can we add some unit tests to test serialization + deserialization?

haiqi96 commented on components/core/src/clp/ir/Constant.hpp:
constants.hpp or Constants.hpp?

kirkrodrigues replied:
Lowercase

haiqi96 commented on components/core/src/clp/ir/Constant.hpp:
```diff
@@ -0,0 +1,8 @@
+#define CLP_IR_CONSTANTS_HPP
+
+namespace clp::ir {
+constexpr char cIrExtension[] = ".clp.zst";
```
string_view doesn't have a "+" operator. in that case, can we change the line 
`string ir_file_name = file_orig_id + .... + ir::cIrFileExtension;` to 
`string ir_file_name = file_orig_id + ... + string(ir::cIrFileExtension);`?

kirkrodrigues replied:
How about we compose the string piece-by-piece?

haiqi96 commented on components/core/src/clp/ir/LogEventSerializer.hpp:
```diff
@@ -0,0 +1,114 @@
+     * Creates a Zstandard compressed IR on the disk, and writes the preamble to the IR.
+     * @param file_path
+     */
+    [[nodiscard]] auto open(std::string const& file_path) -> ErrorCode;
```
if serialize_log_event can return bool, can this method also return bool instead of errorcode?

kirkrodrigues replied:
Sure

haiqi96 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +122,128 @@ bool FileDecompressor::decompress_file(
+        {
+            SPDLOG_ERROR(
+                    "Failed to create directory structure {}, errno={}",
+                    output_dir.c_str(),
```
```suggestion
                    temp_output_dir.c_str(),
```

LinZhihao-723 commented on components/core/src/clp/ir/LogEventSerializer.cpp:
```diff
@@ -0,0 +1,159 @@
+        );
+    }
+
+    if (false == res) {
+        close_writer();
+        return true;
+    }
+
+    m_is_open = true;
+
+    // Flush the preamble
+    flush();
+
+    return false;
```
```suggestion
    if (false == res) {
        close_writer();
        return false;
    }

    m_is_open = true;

    // Flush the preamble
    flush();

    return true;
```
we should return `false` to indicate error right?

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +122,128 @@ bool FileDecompressor::decompress_file(
+            }
+            begin_message_ix = end_message_ix;
+
+            if (auto const error_code = ir_serializer.open(temp_ir_path.string());
```
not error_code

LinZhihao-723 commented on components/core/src/clp/clp/FileDecompressor.cpp:
```diff
@@ -76,4 +122,128 @@ bool FileDecompressor::decompress_file(
+
+    LogEventSerializer<four_byte_encoded_variable_t> ir_serializer;
+    // Open output IR file
+    if (auto const error_code = ir_serializer.open(temp_ir_path.string());
```
not error_code

kirkrodrigues reviewed (approved):
For the PR title, how about:

core-clp: Add support for decompressing a specific file split from a clp archive into one or more IR files.
